File: frontend/server/contexts/user/application/use-cases/create-user.ts
import { DefineUseCase } from "./types";
import { create as _createProfile } from "../../domain/entities/profile";
import { create as _createUser } from "../../domain/entities/user";
import { UserRepository } from "../../domain/repositories/user-repository";
import { defineEmail, definePassword } from "../../domain/value-objects";

interface CreateUser {
  email: string;
  password: string;
  givenName: string;
  middleName: string;
}

export function defineUseCase({ userRepository }: DefineUseCase) {
  return async function useCase(input: CreateUser) {
    const user = _createUser({
      email: defineEmail(input.email),
      password: definePassword(input.email),
      profile: _createProfile({
        givenName: input.givenName,
        middleName: input.middleName,
      }),
    });

    return userRepository.save(user);
  };
}

File: server/internal/user/application/commands/create_user.go
package commands

import (
	"context"

	"github.com/dritelabs/accounts/internal/user/domain/entities"
	"github.com/dritelabs/accounts/internal/user/domain/repositories"
)

type NewCreateUserHandlerConfig struct {
	UserRepository repositories.UserRepository
}

type CreateUserCommand struct {
	Email      string
	GivenName  string
	MiddleName string
	Password   string
}

type CreateUserHandler struct {
	userRepository repositories.UserRepository
}

func (h *CreateUserHandler) Handle(ctx context.Context, cmd CreateUserCommand) error {
	usr, err := entities.NewUser(&entities.NewUserConfig{})

	if err != nil {
		return err
	}

	h.userRepository.Save(ctx, usr)

	// for _, e := range usr.DomainEvents() {

	// }

	return nil
}

func NewCreateUserHandler(userRepository repositories.UserRepository) CreateUserHandler {
	return CreateUserHandler{
		userRepository,
	}
}
